Mathlib/Algebra/Group/Basic.lean

```diff
@@ -221,16 +247,70 @@ end AddMonoid
 
 attribute [local simp] mul_assoc sub_eq_add_neg
 
-section CommMonoid
+section Monoid
+variable [Monoid M] {a b c : M} {m n : ℕ}
+
+@[to_additive boole_nsmul]
+lemma pow_boole (P : Prop) [Decidable P] (a : M) :
+    (a ^ if P then 1 else 0) = if P then a else 1 := by simp only [pow_ite, pow_one, pow_zero]
+#align pow_boole pow_boole
+
+@[to_additive nsmul_add_sub_nsmul]
+lemma pow_mul_pow_sub (a : M) (h : m ≤ n) : a ^ m * a ^ (n - m) = a ^ n := by
+  rw [← pow_add, Nat.add_comm, Nat.sub_add_cancel h]
+#align pow_mul_pow_sub pow_mul_pow_sub
+#align nsmul_add_sub_nsmul nsmul_add_sub_nsmul
+
+@[to_additive sub_nsmul_nsmul_add]
+lemma pow_sub_mul_pow (a : M) (h : m ≤ n) : a ^ (n - m) * a ^ m = a ^ n := by
+  rw [← pow_add, Nat.sub_add_cancel h]
+#align pow_sub_mul_pow pow_sub_mul_pow
+#align sub_nsmul_nsmul_add sub_nsmul_nsmul_add
+
+@[to_additive sub_one_nsmul_add]
+lemma mul_pow_sub_one (hn : n ≠ 0) (a : M) : a * a ^ (n - 1) = a ^ n := by
+  rw [← pow_succ', Nat.sub_add_cancel $ Nat.one_le_iff_ne_zero.2 hn]
+
+@[to_additive add_sub_one_nsmul]
+lemma pow_sub_one_mul (hn : n ≠ 0) (a : M) : a ^ (n - 1) * a = a ^ n := by
+  rw [← pow_succ, Nat.sub_add_cancel $ Nat.one_le_iff_ne_zero.2 hn]
+
+/-- If `x ^ n = 1`, then `x ^ m` is the same as `x ^ (m % n)` -/
+@[to_additive nsmul_eq_mod_nsmul "If `n • x = 0`, then `m • x` is the same as `(m % n) • x`"]
+lemma pow_eq_pow_mod (m : ℕ) (ha : a ^ n = 1) : a ^ m = a ^ (m % n) := by
+  calc
+    a ^ m = a ^ (m % n + n * (m / n)) := by rw [Nat.mod_add_div]
+    _ = a ^ (m % n) := by simp [pow_add, pow_mul, ha]
+#align pow_eq_pow_mod pow_eq_pow_mod
+#align nsmul_eq_mod_nsmul nsmul_eq_mod_nsmul
+
+@[to_additive] lemma pow_mul_pow_eq_one : ∀ n, a * b = 1 → a ^ n * b ^ n = 1
+  | 0, _ => by simp
+  | n + 1, h =>
+    calc
+      a ^ n.succ * b ^ n.succ = a ^ n * a * (b * b ^ n) := by rw [pow_succ, pow_succ']
+      _ = a ^ n * (a * b) * b ^ n := by simp only [mul_assoc]
+      _ = 1 := by simp [h, pow_mul_pow_eq_one]
+#align pow_mul_pow_eq_one pow_mul_pow_eq_one
+#align nsmul_add_nsmul_eq_zero nsmul_add_nsmul_eq_zero
+
+end Monoid
 
-variable {M : Type u} [CommMonoid M] {x y z : M}
+section CommMonoid
+variable [CommMonoid M] {x y z : M}
 
 @[to_additive]
 theorem inv_unique (hy : x * y = 1) (hz : x * z = 1) : y = z :=
   left_inv_eq_right_inv (Trans.trans (mul_comm _ _) hy) hz
 #align inv_unique inv_unique
 #align neg_unique neg_unique
 
+@[to_additive nsmul_add] lemma mul_pow (a b : M) : ∀ n, (a * b) ^ n = a ^ n * b ^ n
+  | 0 => by rw [pow_zero, pow_zero, pow_zero, one_mul]
+  | n + 1 => by rw [pow_succ', pow_succ', pow_succ', mul_pow, mul_mul_mul_comm]
+#align mul_pow mul_pow
+#align nsmul_add nsmul_add
```

Comment:
Maybe this belongs together with `pow_add`/`pow_mul` in `Defs`?

Reply:
Not so easily because it depends on `mul_mul_mul_comm` which is in `Basic`.